src/utils/zip.ts

```diff
@@ -0,0 +1,109 @@
+import { error, warn } from 'electron-log';
+import fs from 'fs';
+import { pathExists } from 'fs-extra';
+import { basename } from 'path';
+import archiver from 'archiver';
+import unzipper from 'unzipper';
+
+/**
+ * Unzips `zip` into `destination`
+ */
+export const unzip = (zip: string, destination: string): Promise<void> => {
+  return new Promise(async (resolve, reject) => {
+    try {
+      const exists = await pathExists(zip);
+      if (!exists) {
+        throw Error(`${zip} does not exist!`);
+      }
+      const stream = fs
+        .createReadStream(zip)
+        .pipe(unzipper.Extract({ path: destination }));
+
+      stream.on('close', resolve);
+      stream.on('error', err => {
+        error(`Could not unzip ${zip} into ${destination}:`, err);
+        reject(err);
+      });
+    } catch (err) {
+      reject(err);
+    }
+  });
+};
+
+interface ZipArgs {
+  /** The destination of the generated zip */
+  destination: string;
+  objects: Array<{
+    /** Object to serialize (with `JSON.stringify`) and store in the zip */
+    object: any;
+    /** Name of this object in the generated zip */
+    name: string;
+  }>;
+  /** Files or folders to include  */
+  paths: string[];
+}
+
+/**
+ * Adds a raw string into the ZIP archive
+ *
+ * @param archive ZIP archive to add the file to
+ * @param content content to add into archive
+ * @param nameInArchive name of file in archive
+ */
+const addStringToZip = (
+  archive: archiver.Archiver,
+  content: string,
+  nameInArchive: string,
+): void => {
+  archive.append(content, { name: nameInArchive });
+  return;
+};
+
+/**
+ * Add the given path to the archive. If it's a file we add it directly, it it is a directory
+ * we recurse over all the files within that directory
+ *
+ * @param archive ZIP archive to add the file to
+ * @param filePath file to add, absolute path
+ */
+const addFileOrDirectoryToZip = async (archive: archiver.Archiver, filePath: string) => {
+  const isDir = await fs.promises.lstat(filePath).then(res => res.isDirectory());
+  if (isDir) {
+    archive.directory(filePath, basename(filePath));
+  } else {
+    archive.file(filePath, { name: basename(filePath) });
+  }
+};
+
+export const zip = ({ destination, objects, paths }: ZipArgs): Promise<void> =>
+  new Promise(async (resolve, reject) => {
+    const output = fs.createWriteStream(destination);
+    const archive = archiver('zip');
+
+    // finished
+    archive.on('finish', () => resolve());
+    archive.on('error', err => {
```

Comment:
I'm having some issues with testing the `error` and `warning` branches here. How would I solve this with a mock?

Reply:
This was a bit tricky because the `archiver` package exposes its main function as the default export. After a little tinkering, I've gotten a couple tests passing which simulate the `error` and `warning` events. 
1. The `jest.mock('archiver`, ...);` code should be placed in a global mock file because it should be mocked in all tests. I just put it all in one file for easier pasting.
1. I mocked `fs` in these tests because your code used that module. I suggest switching to `fs-extra` which already has a global mock in __mocks__. It would just need to be updated with any missing functions declared there, like `createWriteStream`.
```typescript
import fs from 'fs';
import * as archiver from 'archiver';
import { PassThrough } from 'stream';
import { zip } from './zip';

jest.mock('fs');
jest.mock('archiver', () => {
  // TODO: this code should live in __mocks__/archiver.js

  // eslint-disable-next-line @typescript-eslint/no-var-requires
  const { PassThrough } = require('stream');
  let mockStream: any;
  // return a fake stream when "archiver()" is called in the app
  const ctor = function() {
    mockStream = new PassThrough();
    mockStream.file = jest.fn();
    mockStream.directory = jest.fn();
    mockStream.append = jest.fn();
    mockStream.finalize = jest.fn();
    return mockStream;
  };
  // attach a func to emit events on the stream from the tests
  ctor.mockEmit = (event: string, data: any) => mockStream.emit(event, data);

  return ctor;
});

const fsMock = fs as jest.Mocked<typeof fs>;
const archiverMock = archiver as jest.Mocked<any>;

it('should fail if there is an archiver error', async () => {
  fsMock.createWriteStream.mockReturnValueOnce(new PassThrough() as any);

  const promise = zip({
    destination: 'path-to-zip',
    objects: [],
    paths: [],
  });

  // emit an error after a small delay
  const mockError = new Error('test-error');
  setTimeout(() => {
    archiverMock.mockEmit('error', mockError);
  }, 100);

  await expect(promise).rejects.toEqual(mockError);
});

it('should fail if there is an archiver warning', async () => {
  fsMock.createWriteStream.mockReturnValueOnce(new PassThrough() as any);

  const promise = zip({
    destination: 'path-to-zip',
    objects: [],
    paths: [],
  });

  // emit an error after a small delay
  const mockError = new Error('test-warning');
  setTimeout(() => {
    archiverMock.mockEmit('warning', mockError);
  }, 100);

  await expect(promise).rejects.toEqual(mockError);
});
```